### PR TITLE
Add VecDeque::resize_with

### PR DESCRIPTION
**Pull request by scottmcm**
This already exists on `Vec`; I'm just adding it to `VecDeque`.

I wanted to resize a `VecDeque<Vec<T>>` when I didn't know `T: Clone`, so I couldn't use `.resize(n, Vec::new())`.  With this I could do `.resize_with(n, Vec::new)` instead, which doesn't need `T: Clone`.

Tracking issue: https://github.com/rust-lang/rust/issues/41758

### CODE REVIEW

File: src/liballoc/collections/vec_deque.rs [hunk 1]
```diff
@@ -1920,6 +1920,40 @@ impl<T: Clone> VecDeque<T> {
...
+    /// buf.push_back(5);
+    /// buf.push_back(10);
+    /// buf.push_back(15);
+    /// assert_eq!(buf, [5, 10, 15]);
+    ///
+    /// buf.resize_with(2, || unreachable!());
+    /// assert_eq!(buf, [5, 10]);
+    ///
+    /// buf.resize_with(5, Default::default);
+    /// assert_eq!(buf, [5, 10, 0, 0, 0]);
```
Comment (joshtriplett) [inline src/liballoc/collections/vec_deque.rs]: Please include a test with a closure that doesn't return the same result every time, such as `|| { x += 1; x }` (given a preceding `let mut x = 0;`).

Other than that, LGTM.

Comment (scottmcm) [reply src/liballoc/collections/vec_deque.rs]: Makes sense!  Test added.

File: src/liballoc/collections/vec_deque.rs [hunk 2]
```diff
@@ -1920,6 +1920,44 @@ impl<T: Clone> VecDeque<T> {
...
+    ///
+    /// buf.resize_with(2, || unreachable!());
+    /// assert_eq!(buf, [5, 10]);
+    ///
+    /// let mut state = 100;
+    /// buf.resize_with(5, || { state += 1; state });
+    /// assert_eq!(buf, [5, 10, 101, 102, 103]);
+    /// ```
+    #[unstable(feature = "vec_resize_with", issue = "41758")]
+    pub fn resize_with(&mut self, new_len: usize, generator: impl FnMut()->T) {
```
Comment (Centril) [inline src/liballoc/collections/vec_deque.rs]:
```suggestion
    pub fn resize_with(&mut self, new_len: usize, generator: impl FnMut() -> T) {
```

Also, we generally don't use `arg: impl Trait` in the standard library so I think this should be a type param instead.

**Pull request merged**